### PR TITLE
Trainer clean up

### PR DESCRIPTION
**Pull request by jelmervdl**
Restructured the trainer codebase a bit to be less stateful, more functional, better typed, and even added some unit tests. Also implemented async shuffling.

Don't try to read the diff. Just go to trainer.py, scroll all the way to the `__name__ == "__main__"` section, and start from there. It will make much more sense from there.

Todo:
- [x] maybe implement alternative to `shuffle.sh` that doesn't shuffle in memory. For speed that should be C++, but for portability it would be nice to have it in Python inside this project as well.

### CODE REVIEW

File: trainer/trainer.py [hunk 1]
```diff
@@ -1,346 +1,606 @@
...
+        }
+
+    def _load_stage(self, ymldata:dict, stage_name:str, available_datasets:Dict[str,Dataset], seed:int) -> Stage:
+        datasets: List[Tuple[Dataset, float]] = []
+
+        for line in ymldata[stage_name][:-1]:
+            dataset_name, weight = line.split()
+            datasets.append((available_datasets[dataset_name], float(weight)))
+
+        _, dataset_name, max_epochs = ymldata[stage_name][-1].split()
```
Comment (XapaJIaMnu) [inline trainer/trainer.py:346]: `dataset_name` is  a bit misleading here... Should probably be `until_dataset`.

File: trainer/trainer.py [hunk 2]
```diff
@@ -1,346 +1,606 @@
...
+        ```
+        """
+        return [
+            Modifier(name, float(ymldata[name]))
+            for name in MODIFIERS.keys()
+            if name in ymldata
+        ]
+
+
+class CurriculumV2Loader(CurriculumV1Loader):
```
Comment (XapaJIaMnu) [inline trainer/trainer.py:370]: Add a test and example config for this.

File: trainer/trainer.py [hunk 3]
```diff
@@ -1,346 +1,606 @@
...
+        name, frequency = line.split()
+        return Modifier(name, float(frequency))
+
+
+class CurriculumLoader:
+    """Reads curriculum yaml files. Wrapper that decides which reader to use
+    based on a version number that may be in there.
+    """
+
+    IMPLEMENTATIONS={
```
Comment (XapaJIaMnu) [inline trainer/trainer.py:414]: We haven't released it so maybe we can make this a possibility, although the way i imagine it, i don't see a reason why we should have multiple unmerged files per stage...?

File: trainer/trainer.py [hunk 4]
```diff
@@ -1,346 +1,606 @@
...
+        self.epoch_offset = reader.epoch
+        self.line_offset = reader.line
+
+    @property
+    def epoch(self):
+        epoch = self.reader.epoch - self.epoch_offset
+
+        # ... but if the reader is behind on where it was, it hasn't completed
+        # a full epoch yet.
+        if self.reader.line < self.line_offset:
```
Comment (XapaJIaMnu) [inline trainer/trainer.py:443]: I am not sure I understand why this? How/why would epoch move forward without further reading?

File: trainer/trainer.py [hunk 5]
```diff
@@ -1,346 +1,606 @@
...
+    args = parser.parse_args()
+
+    with open(args.config, 'r', encoding='utf-8') as fh:
+        config = yaml.safe_load(fh)
+
+    curriculum = CurriculumLoader().load(config, basepath=os.path.dirname(args.config))
+
+    trainer = Trainer(curriculum, reader=DatasetReader if args.sync else AsyncDatasetReader, flip=args.flip)
+
+    state_harness = StateTracker(args.state or f'{args.config}.state')
```
Comment (XapaJIaMnu) [inline trainer/trainer.py:587]: The state harness variable name... I am not sure I agree here ;d

File: trainer/trainer.py [hunk 6]
```diff
@@ -1,346 +1,606 @@
...
+
+    # Disable resume functionality if we don't want it.
+    if args.do_not_resume:
+        state_harness._restore = lambda trainer: None
```
Comment (XapaJIaMnu) [inline trainer/trainer.py:591]: Is hot patching the function here really the best idea?

**Review by XapaJIaMnu — CHANGES_REQUESTED** [review]
Looks good to me. Most points are minor.

One major point:

What is the purpose of API v2. Technically we haven't released this so we don't need to specify it. We should indeed have an API string in the yml. We should discuss the v2 api offline.

File: trainer/trainer.py [hunk 7]
```diff
@@ -1,346 +1,605 @@
...
 
-    def set_max_epoch(self, new_max_epoch):
-        '''Used for when we want to switch the sampling strategy based on our schedule'''
-        self.max_epoch = new_max_epoch
 
-    def reset_epoch(self):
-        '''Used to reset the training epoch of the file, so that training can continue
-        from the same shuffling point without eextra fluff'''
+class DatasetReader:
+    """Repeats, shuffles and reads a dataset ad infinitum."""
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply trainer/trainer.py:107]: I do not attribute those properties to this python code. If it really does grow tired of it, it can decide to stop generating on its own.

File: trainer/trainer.py (hunk 4 above)
Comment (jelmervdl) [reply trainer/trainer.py:473]: Because at the moment the dataset readers are not reset between stages. If only half of your `dirty` dataset was read during the `start` stage, then in the `mid` stage the rest will be shown first.

The alternative would be to always reset and reshuffle datasets for each stage. That could work if every stage has its own random seed (and one that doesn't overlap with re-iterated datasets, so can't just do `seed+1`) or if we're happy with skipping remainder of datasets at the end of each stage i.e. always start DatasetReader at `prev_stage.epoch + 1` even if it didn't do a full epoch during the last stage.

**Comment by jelmervdl** [comment]
The curriculum v2 and embedded deduplication are all related to me trying to remove the need for a step between having the output from the cleaning and categorised your datasets, and having the input data for your training.

My idea was that the amount of effort necessary for adding another dataset should be pretty minimal. With the above implementation you can just add some more filenames to your yaml and it will just work. The current implementation is not perfect, i.e. if a sentence pair occurs in both _clean_ and _dirty_ and both datasets are used in a stage, those pairs could be seen more often. But at least they're guaranteed to only occur once in each dataset.

I'm also open for adding something more advanced, like the score based deduplication discussed in #41, to the end of the cleaning pipeline. The nasty bit about that is that you have to rescore & regenerate your training datasets. By integrating it into the cleaning pipeline we could keep the scores cached though.

But we do have to offer something to abridge going from the cleaned datasets to the trainer input. We shouldn't expect our end-users to dive into bash and concat + dedupe files themselves.

File: trainer/trainer.py (hunk 4 above)
Comment (XapaJIaMnu) [reply trainer/trainer.py:473]: Oh yeah, i understand. In my original implementation i was not taking into account that i can have a smaller first epoch due to it.

**Comment by jelmervdl** [comment]
I reworked `shuffle.py` a bit. On sigyn I tested it on a bit of data and it shuffled it in 7 minutes (was 17 with the previous implementation.) For comparison, doing it all in memory and without Python, e.g. `pigz -cd *.gz | shuf | pigz -c > shuffled.gz` does it in 5 minutes. But during lunch I've decided that trying to optimise this further in Global Interpreter Locked Python (as opposed to, say, C++) is silly.

File: trainer/trainer.py [hunk 8]
```diff
@@ -648,6 +648,7 @@ def run(self, trainer:Trainer, *args, **kwargs):
 
     model_trainer = subprocess.Popen(
         args.trainer or config['trainer'],
+        bufsize=0, # no limbo for lines written but not yet read by marian
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply trainer/trainer.py:669]: I'm testing with it. Right now it doesn't seem to have any positive or negative impact (marian still reads a lot ahead anyway) so I'm going to remove it again.

File: trainer/trainer.py (hunk 3 above)
Comment (jelmervdl) [reply trainer/trainer.py:407]: Collapsed it all into a version 1. I've left the IMPLEMENTATIONS switch there for future versions. Modifiers will need a rework some time in the future when there are more options, but I don't have the best ideas for that right now.

File: trainer/trainer.py [hunk 9]
```diff
@@ -1,346 +1,605 @@
...
+
+    def close(self):
+        if self._fh:
+            self._fh.close()
+
+    def _open(self):
+        # Open temporary file which will contain shuffled version of `cat self.files`
+        fh = TemporaryFile(mode='w+', encoding='utf-8')
+
+        # Shuffle data to the temporary file. Deduplicate on the src side
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply trainer/trainer.py]: Deduplication has been removed. It will have to happen somewhere earlier in the pipeline. Also support for multiple files has been removed from the configuration file to force people to do their deduplication & concatenation beforehand.
